core/src/main/java/hudson/model/Actionable.java

```diff
@@ -121,30 +134,150 @@
+        if (a == null) {
+            return false;
+        }
+        // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
+        return getOrCreateActions().removeAll(Collections.singleton(a));
+    }
+
+    /**
+     * Removes any actions of the specified type.
+     * Note: calls to {@link #getAllActions()} that happen before calls to this method may not see the update.
+     * Note: this method does not affect transient actions contributed by a {@link TransientActionFactory}
+     *
+     * @param clazz the type of actions to remove
+     * @return {@code true} if this actions changed as a result of the call
+     * @since FIXME
+     */
+    @SuppressWarnings({"ConstantConditions","deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean removeActions(@Nonnull Class<? extends Action> clazz) {
+        if (clazz == null) {
+            throw new IllegalArgumentException("Action type must be non-null");
+        }
+        // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
+        List<Action> old = new ArrayList<Action>();
+        List<Action> current = getOrCreateActions();
+        for (Action a : current) {
+            if (clazz.isInstance(a)) {
+                old.add(a);
+            }
+        }
+        return current.removeAll(old);
+    }
+
+    /**
+     * Replaces any actions of the specified type by the supplied action.
+     * Note: calls to {@link #getAllActions()} that happen before calls to this method may not see the update.
+     * Note: this method does not affect transient actions contributed by a {@link TransientActionFactory}
+     *
+     * @param clazz the type of actions to replace (note that the action you are replacing this with need not extend
+     *              this class)
+     * @param a     the action to replace with
+     * @return {@code true} if this actions changed as a result of the call
+     * @since FIXME
+     */
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
+        if (clazz == null) {
+            throw new IllegalArgumentException("Action type must be non-null");
+        }
+        if (a == null) {
+            throw new IllegalArgumentException("Action must be non-null");
+        }
+        // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
+        List<Action> old = new ArrayList<Action>();
+        List<Action> current = getOrCreateActions();
+        boolean found = false;
+        for (Action a1 : current) {
+            if (!found && a.equals(a1)) {
+                found = true;
+            } else if (clazz.isInstance(a1) && !a.equals(a1)) {
+                old.add(a1);
+            }
+        }
+        current.removeAll(old);
+        if (!found) {
```

Comment:
Basically, we run through the actions, if we found the action in the list already, either you didn't override `equals` in which case it is == and therefore there is no point in removing the action just to add it again... or you did override `equals` in which case you have done that correctly and thus we would just be removing one equivalent object to replace with another.